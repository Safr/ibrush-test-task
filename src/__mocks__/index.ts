export const orderBook = {
  BTC_USD: {
    ask: ['6891', '0.005', '34.455'],
    ask_amount: '2413543.16939829',
    ask_quantity: '190.96765367',
    ask_top: '6887.20157116',
    bid: ['6849.7394927', '0.75', '5137.30461952'],
    bid_amount: '286086.48310063',
    bid_quantity: '11524.76654836',
    bid_top: '6875',
  },
};

export const order = {
  avg: '6875.86448007',
  buy_price: '6859.227881',
  high: '6969.54201841',
  last_trade: '6864.01',
  low: '6800',
  sell_price: '6873.71642794',
  updated: 1528752429,
  vol: '1112.01492053',
  vol_curr: '7632881.53471028',
};

export const currentOrder = {
  BTC_USD: {
    avg: '6875.86448007',
    buy_price: '6859.227881',
    high: '6969.54201841',
    last_trade: '6864.01',
    low: '6800',
    sell_price: '6873.71642794',
    updated: 1528752429,
    vol: '1112.01492053',
    vol_curr: '7632881.53471028',
  },
};

export const orders = { currentOrder };
export const changedOrders = [''];
export const orderBookInfo = [['5897.8242', '0.003', '17.6934726']];

// export const fetchOrders = () => Promise.resolve({
//   orders: {
//     BTC_USD: {
//       avg: '6875.86448007',
//       buy_price: '6859.227881',
//       high: '6969.54201841',
//       last_trade: '6864.01',
//       low: '6800',
//       sell_price: '6873.71642794',
//       updated: 1528752429,
//       vol: '1112.01492053',
//       vol_curr: '7632881.53471028',
//     },
//   },
// });
