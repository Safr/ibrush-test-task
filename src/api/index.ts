export const fetchOrders = async (): Promise<object|string> => {
// export const fetchOrders = async () => {
  const endPoint = 'https://api.exmo.com/v1/ticker/';
  try {
    const apiResponse = await fetch(endPoint);
    const orders = await apiResponse.json();
    return orders;
  } catch (e) {
    return e.message;
  }
};

export const fetchOrderBook = async (pair: string): Promise<object|string> => {
  const endPoint = `https://api.exmo.com/v1/order_book/?limit=50&pair=${pair}`;
  try {
    const apiResponse = await fetch(endPoint);
    const orderBookInfo = await apiResponse.json();
    return orderBookInfo;
  } catch (e) {
    return e.message;
  }
};
