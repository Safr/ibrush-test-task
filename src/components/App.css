.Container {
  display: flex;
  align-items: center;
  flex-direction: column;
  max-width: 980px;
  margin: 0 auto;
  padding: 10px;
  padding-top: 100px;
}

@media screen and (max-width: 660px) {
  .Container {
    padding-top: 50px;
  }
}