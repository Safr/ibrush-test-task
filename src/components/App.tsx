import React from 'react';
import classes from './App.css';
import Header from './Header/Header';
import OrdersWrapper from './OrdersWrapper/OrdersWrapper';
import OrderButton from './OrderButton/OrderButton';
import Popup from './Popup/Popup';
import { fetchOrders, fetchOrderBook } from '../api';
import { OrderInterface } from '../types';
import { handleErrors } from '../helpers';

interface AppState {
  orders: null | string | object;
  changedOrders: null | string[];
  currentOrder: null | object;
  orderBook: null | object;
  isFetching: boolean;
  isOpen: boolean;
  error: string;
}

class App extends React.Component<{}, AppState> {
  state = {
    orders: null,
    changedOrders: [],
    currentOrder: null,
    orderBook: null,
    isFetching: false,
    isOpen: false,
    error: '',
  };
  componentDidMount() {
    const orders = this.getOrdersFromStorage();
    if (orders !== null) {
      this.setState({ orders });
    }
  }
  getOrdersFromStorage = () => {
    let orders = localStorage.getItem('orders');
    if (orders !== null) {
      orders = JSON.parse(orders);
      return orders;
    }
    return null;
  }
  handleOpenPopup = (orderName: string, className: string, order: OrderInterface) => {
    this.setState({
      currentOrder: null,
      orderBook: null,
    });
    this.handleOrderBook(orderName);
    this.setState({
      isOpen: true,
      currentOrder: { [orderName]: order, className },
    });
  }
  handleClosePopup = () => {
    this.setState({
      isOpen: false,
    });
  }
  handleOrderBook = async (pair: string) => {
    const orderBook = await fetchOrderBook(pair);
    if (handleErrors(orderBook)) {
      this.setState({
        error: 'Возникли неполадки на стороне сервера:(',
      });
    }
    this.setState({ orderBook: { [pair]: orderBook[pair] } });
  }

  handleOrders = async () => {
    const ordersFromStorage = this.getOrdersFromStorage();
    this.setState({ isFetching: true });
    const orders = await fetchOrders();
    if (handleErrors(orders)) {
      this.setState({
        isFetching: false,
        error: 'Возникли неполадки на стороне сервера:(',
      });
    } else {
      localStorage.setItem('orders', JSON.stringify(orders));
      const changedOrders = ordersFromStorage ? this.compareOrders(orders as object, ordersFromStorage) : null;
      this.setState({
        orders,
        changedOrders,
        isFetching: false,
      });
    }
  }

  compareOrders = (orders: object, ordersFromStorage: string | object) => {
    const orderItems = Object.keys(orders).slice(0, 10).map(orderName => orders[orderName]);
    const orderItemsFromStorage = Object.keys(ordersFromStorage).slice(0, 10).map(orderName => orders[orderName]);
    const changedOrders = orderItems.map((orderItem, idx) => {
      const orderStatePrice = orderItem.buy_price;
      const orderStoragePrice = orderItemsFromStorage[idx].buy_price;
      if (orderStatePrice > orderStoragePrice) {
        return 'up';
      }
      if (orderStatePrice < orderStoragePrice) {
        return 'down';
      }
      return '';
    });
    return changedOrders;

  }
  render() {
    const { orders, changedOrders, currentOrder, orderBook, isFetching, isOpen, error } = this.state;
    return (
      <>
      <Header />
      <div className={classes.Container}>
        <OrdersWrapper
          orders={orders}
          changedOrders={changedOrders}
          isFetching={isFetching}
          onOpenPopup={this.handleOpenPopup}
          error={error}
        />
        <OrderButton onFetchOrders={this.handleOrders} />
      </div>
      <Popup
        onClosePopup={this.handleClosePopup}
        isOpen={isOpen}
        currentOrder={currentOrder}
        orderBook={orderBook && orderBook}
        error={error}
      />
      </>
    );
  }
}

export default App;
