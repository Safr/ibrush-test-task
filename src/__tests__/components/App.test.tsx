import React from 'react';
import { shallow } from 'enzyme';
import App from '../../components/App';
import { fetchOrders, fetchOrderBook } from '../../api';
import { order } from '../../__mocks__';

describe('App Component', () => {
  let wrapper;
  let fetchSpy;
  let fetchPromise;
  beforeEach(() => {
    wrapper = shallow(
      <App />,
    );
    fetchSpy = global.fetch;
    fetchPromise = Promise.resolve({
      json: () => Promise.resolve({}),
    });
    global.fetch = () => fetchPromise;
  });
  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  test('should set state from localStorage', () => {
    const KEY = 'orders';
    const VALUE = 'bar';
    wrapper.instance().componentDidMount();
    wrapper.update();
    expect(localStorage.getItem).toHaveBeenLastCalledWith(KEY);
  });
  test('should set state isOpen to false', () => {
    wrapper.instance().handleClosePopup();
    expect(wrapper.state('isOpen')).toBe(false);
  });
  test('should handle handleOrderBook', async () => {
    wrapper.instance().handleOrderBook();
    await expect(fetchOrderBook('BTC_USD')).resolves.toEqual({});
    expect(wrapper.state('orderBook')).toEqual({});
  });
  test('should handle handleOrders', async () => {
    wrapper.instance().handleOrders();
    await expect(fetchOrders()).resolves.toEqual({});
    expect(wrapper.state('orders')).toEqual({});
  });
  test('should handle openPopup', () => {
    const expected = {
      BTC_USD: order,
      className: '',
    };
    wrapper.instance().handleOpenPopup('BTC_USD', '', order);
    expect(wrapper.state('isOpen')).toBe(true);
    expect(wrapper.state('currentOrder')).toEqual(expected);
    expect(wrapper.state('orderBook')).toBe(null);
  });
});
