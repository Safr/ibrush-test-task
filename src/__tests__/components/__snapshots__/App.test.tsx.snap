// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should render correctly 1`] = `
<React.Fragment>
  <Header />
  <div>
    <OrdersWrapper
      changedOrders={Array []}
      error=""
      isFetching={false}
      onOpenPopup={[Function]}
      orders={null}
    />
    <OrderButton
      onFetchOrders={[Function]}
    />
  </div>
  <Popup
    currentOrder={null}
    error=""
    isOpen={false}
    onClosePopup={[Function]}
    orderBook={null}
  />
</React.Fragment>
`;
